### PR TITLE
Enhance the default configurations for TiFlash

### LINKED ISSUE #4238: Enhance the default configurations for TiFlash
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

We have set the default values for all the configurations for TiFlash in TiDB Operator, however, TiFlash will set default values by itself in v5.4.0, so after that, we do not need to set the default values in TiDB Operator anymore.
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Close #4238
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 

    - [x] Deploy v5.3.0 cluster and check the config of TiFlash
    - [x] Deploy v5.4.0-pre with v1.2.4 operator and check the config of TiFlash
    - [x] Deploy v5.4.0-pre with v1.2.4 operator and set default config, then upgrade operator. Check that there is no rolling upgrade
    - [x] Deploy v5.4.0-pre cluster and check the config of TiFlash
    - [x] Deploy v5.4.0-pre cluster with TlS-enabled and check the config of TiFlash

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Remove and change some default configurations for TiFlash due to config change about v5.4.0 TiFlash. If deploy v5.4 tidb cluster by v1.2 operator, upgrade operator will cause rolling upgrade of tiflash
```

### CODE REVIEW

File: pkg/manager/member/tiflash_util.go [hunk 1]
```diff
@@ -99,6 +106,119 @@ func buildSidecarContainer(name, path, image string,
...
+		// check "path" to be compatible with old version
+		if common.Get("path") == nil && common.Get("storage") == nil {
+			paths := []string{}
+			for i := range tc.Spec.TiFlash.StorageClaims {
+				paths = append(paths, fmt.Sprintf("/data%d/db", i))
+			}
+			if len(paths) == 0 {
+				paths = []string{"/data0/db"}
+			}
+			common.Set("storage.main.dir", paths)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_util.go:150]: This is separated by `,` or ` `?

File: pkg/util/cmpver/cmpver.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+	c, err := NewConstraint(op, ver2)
+	if err != nil {
+		return false, err
+	}
+
+	return c.Check(ver1)
+}
+
+// Compare compare ver1 and ver2
```
Comment (DanielZhangQD) [inline pkg/util/cmpver/cmpver.go:54]:
```suggestion
// CompareByStr compares ver1 and ver2
```

File: pkg/manager/member/tiflash_util.go (hunk 1 above)
Comment (KanShiori) [reply pkg/manager/member/tiflash_util.go:150]: it's a slice.

File: pkg/manager/member/tiflash_util.go [hunk 3]
```diff
@@ -99,6 +106,119 @@ func buildSidecarContainer(name, path, image string,
...
+		pdAddr := fmt.Sprintf("%s.%s.svc:2379", controller.PDMemberName(name), ns)
+		if len(clusterDomain) > 0 {
+			pdAddr = "PD_ADDR"
+		} else if noLocalPD {
+			pdAddr = fmt.Sprintf("%s.%s.svc%s:2379", controller.PDMemberName(ref.Name), ref.Namespace, controller.FormatClusterDomain(ref.ClusterDomain))
+		}
+		common.SetIfNil("raft.pd_addr", pdAddr)
+
+		// server
+		common.SetIfNil("server.engine-addr", fmt.Sprintf("%s-POD_NUM.%s.%s.svc%s:3930", controller.TiFlashMemberName(name), controller.TiFlashPeerMemberName(name), ns, controller.FormatClusterDomain(clusterDomain)))
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_util.go:189]: L197 should be enough?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by july2993 — APPROVED** [review]

**Pull request merged**